### PR TITLE
Refactor syncer

### PR DESCRIPTION
**Pull request by eseliger**
This does a lot of changes that I found on the way to making the syncer use a deterministic token. It should be reviewed commit by commit.

### CODE REVIEW

**Comment by eseliger** [comment]
TODO: 2 external services might have the same token associated

File: internal/campaigns/changeset.go [hunk 1]
```diff
@@ -378,8 +379,28 @@ func (c *Changeset) URL() (s string, err error) {
...
+// Dedupe deduplicates events per changeset based on their Kind+Key to avoid
+// conflicts when inserting into database.
+func (ce ChangesetEvents) Dedupe() []*ChangesetEvent {
+	uniqueEvents := make(map[string]struct{}, len(ce))
+	var events []*ChangesetEvent
+
+	for _, e := range ce {
+		k := string(e.Kind) + e.Key
+		if _, ok := uniqueEvents[k]; ok {
+			log15.Info("dropping duplicate changeset event", "changeset_id", e.ChangesetID, "kind", e.Kind, "key", e.Key)
```
Comment (LawnGnome) [inline internal/campaigns/changeset.go]: Does this need to be `Info`? How common do we think this is?

File: enterprise/internal/campaigns/syncer.go [hunk 2]
```diff
@@ -88,13 +88,20 @@ func (s *SyncRegistry) Add(kind, config string) {
 		return
 	}
 
+	sourcer := repos.NewSourcer(s.HTTPFactory)
+	css, err := buildChangesetSource(s.Ctx, s.RepoStore, sourcer, extSvc)
```
Comment (LawnGnome) [inline enterprise/internal/campaigns/syncer.go]: `css` might be a confusing name here, given its other meaning. :smile:

**Review by LawnGnome — APPROVED** [review]
A good set of cleanups here!

File: enterprise/internal/campaigns/syncer.go [hunk 3]
```diff
@@ -181,19 +172,40 @@ func (s *SyncRegistry) EnqueueChangesetSyncs(ctx context.Context, ids []int64) e
...
+		if es.LastSyncAt != nil {
+			extSvc.LastSyncAt = *es.LastSyncAt
+		}
+		if es.NextSyncAt != nil {
+			extSvc.NextSyncAt = *es.NextSyncAt
+		}
+		if es.NamespaceUserID != nil {
+			extSvc.NamespaceUserID = *es.NamespaceUserID
+		}
+		s.Add(extSvc)
```
Comment (mrnugget) [inline enterprise/internal/campaigns/syncer.go:208]: Nit: I think the conversion could be moved into a separate helper function.

File: internal/campaigns/changeset_event.go [hunk 4]
```diff
@@ -78,6 +78,26 @@ const (
 	ChangesetEventKindGitLabUnmarkWorkInProgress ChangesetEventKind = "gitlab:unmark_wip"
 )
 
+type ChangesetEvents []*ChangesetEvent
+
+// Dedupe deduplicates events per changeset based on their Kind+Key to avoid
+// conflicts when inserting into database.
+func (ce ChangesetEvents) Dedupe() []*ChangesetEvent {
```
Comment (mrnugget) [inline internal/campaigns/changeset_event.go:85]: Is there anything that speaks against adding this dedupe logic to `Changeset.Events()`?

**Review by mrnugget — APPROVED** [review]

File: internal/campaigns/changeset.go (hunk 1 above)
Comment (eseliger) [reply internal/campaigns/changeset.go]: It should ideally never happen I think. It usually means our key is not actually unique, or there is some weird state where something is in the timeline twice(?) I think it'd leave it as is -for now- and see how frequent it happens. It's been there before so I think if it hasn't annoyed any of us yet, it's still ok.

**Comment by LawnGnome** [comment]
The latest changes LGTM. :+1:

**Pull request merged**